File: agent/src/limits.ts
// Bridge limits: the single source of truth loaded from schemas/bridge_limits.json.
//
// Mirror of server/hephaestus/agent_bridge/limits.py. Every architecture §5
// numeric limit is read from the JSON file at import time; NO limit literal is
// duplicated here. Also exports the bounded validators the sidecar runs before
// trusting a payload: JSON structural caps, x-hephaestus-maxUtf8Bytes enforcement
// (with unpaired-surrogate rejection — never replacement-character coercion), and
// a PNG/JPEG header parser that recovers dimensions before any full decode.

import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

export interface DelegationLimits {
  readonly deadline_default_seconds: number;
  readonly deadline_min_seconds: number;
  readonly deadline_max_seconds: number;
  readonly grace_seconds: number;
}

export interface BridgeLimits {
  readonly version: number;
  readonly wire: { readonly frame_version: number; readonly max_frame_bytes: number };
  readonly json: {
    readonly max_depth: number;
    readonly max_members: number;
    readonly max_array_items: number;
    readonly max_string_bytes: number;
  };
  readonly binary: { readonly max_binary_bytes: number };
  readonly image: {
    readonly max_image_bytes: number;
    readonly max_width: number;
    readonly max_height: number;
    readonly max_total_pixels: number;
    readonly max_images_per_result: number;
  };
  readonly rpc: { readonly max_pending: number; readonly py_handler_workers: number };
  readonly admission: { readonly run_slots: number };
  readonly events: { readonly buffered_events: number };
  readonly timeouts: {
    readonly tool_seconds: number;
    readonly cad_build_seconds: number;
    readonly delegation: DelegationLimits;
  };
  readonly prompt: { readonly max_utf8_bytes: number };
  readonly text_result: { readonly max_bytes: number; readonly max_lines: number };
}

// `schemas/bridge_limits.json` is the single source of truth for these bounds:
// Python reads the same file, so a limit can never be raised on one side of the
// bridge alone. That makes *where* it is on disk layout-dependent, and there are
// two layouts:
//
//   - source / `tsc` output: agent/src/limits.ts and agent/dist/limits.js both
//     sit two levels under the repo root, so `../../schemas/` finds it;
//   - the bundled sidecar shipped in the wheel: there is no repo. The bundler
//     copies the schema to `schemas/bridge_limits.json` beside the entry point,
//     where it is covered by the sidecar's integrity manifest — so a tampered
//     limits file is a fail-closed refusal, not a silently widened bound.
//
// Candidates are tried in order and the first that exists wins. An explicit
// HEPHAESTUS_BRIDGE_LIMITS still overrides both (tests pin bounds with it).
const LIMITS_CANDIDATES = [
  new URL("../../schemas/bridge_limits.json", import.meta.url),
  new URL("./schemas/bridge_limits.json", import.meta.url),
];

function loadLimits(): BridgeLimits {
  const override = process.env.HEPHAESTUS_BRIDGE_LIMITS;
  if (override !== undefined) {
    return JSON.parse(readFileSync(override, "utf8")) as BridgeLimits;
  }
  const tried: string[] = [];
  for (const candidate of LIMITS_CANDIDATES) {
    const path = fileURLToPath(candidate);
    tried.push(path);
    if (existsSync(path)) {
      return JSON.parse(readFileSync(path, "utf8")) as BridgeLimits;
    }
  }
  throw new Error(
    `bridge_limits.json not found (tried: ${tried.join(", ")}); ` +
      "set HEPHAESTUS_BRIDGE_LIMITS to name it explicitly",
  );
}

export const LIMITS: BridgeLimits = loadLimits();

export const WIRE_FRAME_VERSION = LIMITS.wire.frame_version;
export const WIRE_MAX_FRAME_BYTES = LIMITS.wire.max_frame_bytes;
export const MAX_JSON_DEPTH = LIMITS.json.max_depth;
export const MAX_JSON_MEMBERS = LIMITS.json.max_members;
export const MAX_JSON_ARRAY_ITEMS = LIMITS.json.max_array_items;
export const MAX_STRING_BYTES = LIMITS.json.max_string_bytes;
export const MAX_BINARY_BYTES = LIMITS.binary.max_binary_bytes;
export const MAX_IMAGE_BYTES = LIMITS.image.max_image_bytes;
export const MAX_IMAGE_WIDTH = LIMITS.image.max_width;
export const MAX_IMAGE_HEIGHT = LIMITS.image.max_height;
export const MAX_TOTAL_PIXELS = LIMITS.image.max_total_pixels;
export const MAX_IMAGES_PER_RESULT = LIMITS.image.max_images_per_result;
export const MAX_PENDING_RPC = LIMITS.rpc.max_pending;
export const PROMPT_MAX_UTF8_BYTES = LIMITS.prompt.max_utf8_bytes;
export const BUFFERED_EVENTS_MAX = LIMITS.events.buffered_events;

// The three outbound deadlines, in milliseconds (audit-2026-09-04
// J-http-limits-11). The shared document states them in SECONDS, because that
// is the unit every other consumer reads; the conversion happens here, once, so
// no caller multiplies by a thousand of its own.
//
// TOOL_TIMEOUT_MS is the peer's default and the deadline of an ordinary tool
// call. CAD_BUILD_TIMEOUT_MS is the class for the tools that run the sandboxed
// worker; it was declared on the Python supervisor, where a build never travels
// (a build goes sidecar->Python), so no call site could ever select it and the
// 300 s three documents state as fact was unreachable — every tool died at the
// peer's hardcoded 120 s. DELEGATION_GRACE_MS is the margin added to a
// synchronous delegation's own deadline so the delegation's `timed_out`
// terminal wins over a transport timeout instead of racing it.
export const TOOL_TIMEOUT_MS = LIMITS.timeouts.tool_seconds * 1000;
export const CAD_BUILD_TIMEOUT_MS = LIMITS.timeouts.cad_build_seconds * 1000;
export const DELEGATION_GRACE_MS = LIMITS.timeouts.delegation.grace_seconds * 1000;

export class LimitError extends Error {
  constructor(
    readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "LimitError";
  }
}

export class ImageError extends LimitError {
  constructor(code: string, message: string) {
    super(code, message);
    this.name = "ImageError";
  }
}

export interface ImageDims {
  readonly width: number;
  readonly height: number;
  readonly kind: "png" | "jpeg";
}

type JsonNode = null | boolean | number | string | JsonNode[] | { [k: string]: JsonNode };

/** Reject well-formed-but-pathological JSON against the §5 structural caps. */
export function validateJsonStructure(value: JsonNode): void {
  const stack: Array<{ node: JsonNode; depth: number }> = [{ node: value, depth: 1 }];
  while (stack.length > 0) {
    const { node, depth } = stack.pop() as { node: JsonNode; depth: number };
    if (depth > MAX_JSON_DEPTH) {
      throw new LimitError("json_too_deep", `nesting exceeds depth ${MAX_JSON_DEPTH}`);
    }
    if (Array.isArray(node)) {
      if (node.length > MAX_JSON_ARRAY_ITEMS) {
        throw new LimitError(
          "json_array_too_long",
          `array has ${node.length} items (max ${MAX_JSON_ARRAY_ITEMS})`,
        );
      }
      for (const child of node) stack.push({ node: child, depth: depth + 1 });
    } else if (node !== null && typeof node === "object") {
      const keys = Object.keys(node);
      if (keys.length > MAX_JSON_MEMBERS) {
        throw new LimitError(
          "json_too_many_members",
          `object has ${keys.length} members (max ${MAX_JSON_MEMBERS})`,
        );
      }
      for (const key of keys) {
        checkStringBytes(key);
        stack.push({ node: node[key] as JsonNode, depth: depth + 1 });
      }
    } else if (typeof node === "string") {
      checkStringBytes(node);
    }
  }
}

function checkStringBytes(s: string): void {
  const n = utf8LenStrict(s);
  if (n > MAX_STRING_BYTES) {
    throw new LimitError("json_string_too_large", `string is ${n} bytes (max ${MAX_STRING_BYTES})`);
  }
}

/**
 * Exact UTF-8 byte length, rejecting any unpaired UTF-16 surrogate as
 * `invalid_unicode_scalar`. Because JavaScript must not silently coerce lone
 * surrogates to U+FFFD, we validate pairing first, then measure — so
 * Buffer.byteLength only ever sees a well-formed string.
 */
export function utf8LenStrict(s: string): number {
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c >= 0xd800 && c <= 0xdbff) {
      const next = i + 1 < s.length ? s.charCodeAt(i + 1) : 0;
      if (next < 0xdc00 || next > 0xdfff) {
        throw new LimitError(
          "invalid_unicode_scalar",
          `unpaired high surrogate U+${c.toString(16).toUpperCase()} at index ${i}`,
        );
      }
      i++; // consume the valid low surrogate
    } else if (c >= 0xdc00 && c <= 0xdfff) {
      throw new LimitError(
        "invalid_unicode_scalar",
        `unpaired low surrogate U+${c.toString(16).toUpperCase()} at index ${i}`,
      );
    }
  }
  return Buffer.byteLength(s, "utf8");
}

/**
 * Enforce the AGGREGATE binary budget of one tool result (J-http-limits-9).
 *
 * `binary.max_binary_bytes` was declared on both sides and validated by
 * nothing: a grep for it returned the Python export, the TypeScript export and
 * the test that pinned it as dead. It has one defensible subject on this
 * bridge — the only binary payloads that cross it are the base64 image blocks
 * of a tool result, and eight megabytes (`max_image_bytes`) times four
 * (`max_images_per_result`) is exactly this cap, so the intended meaning is the
 * per-result *aggregate*. Nothing summed them, so a result with four maximal
 * images was bounded only by the per-image check passing four times, and any
 * future non-image payload had no bound at all.
 *
 * Enforced on BOTH sides (`limits.py`'s `enforce_binary_budget`), because a
 * limit enforced on one side of this bridge is not enforced.
 */
export function enforceBinaryBudget(totalBytes: number, field = "result"): void {
  if (totalBytes > MAX_BINARY_BYTES) {
    throw new LimitError(
      "binary_too_large",
      `${field} carries ${totalBytes} binary bytes (max ${MAX_BINARY_BYTES})`,
    );
  }
}

/** Enforce x-hephaestus-maxUtf8Bytes; return the exact UTF-8 byte length. */
export function enforceMaxUtf8Bytes(value: string, maxBytes: number, field = "value"): number {
  const n = utf8LenStrict(value);
  if (n > maxBytes) {
    throw new LimitError("prompt_too_large", `${field} is ${n} UTF-8 bytes (max ${maxBytes})`);
  }
  return n;
}

const PNG_SIGNATURE = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
// JPEG SOF markers carrying dimensions (exclude DHT/JPG/DAC/RSTn).
const JPEG_SOF = new Set([
  0xc0, 0xc1, 0xc2, 0xc3, 0xc5, 0xc6, 0xc7, 0xc9, 0xca, 0xcb, 0xcd, 0xce, 0xcf,
]);

/** Recover image dimensions from a PNG/JPEG header without a full decode. */
export function parseImageHeader(data: Buffer): ImageDims {
  if (data.length > MAX_IMAGE_BYTES) {
    throw new ImageError("image_too_large", `image is ${data.length} bytes (max ${MAX_IMAGE_BYTES})`);
  }
  let dims: ImageDims;
  if (data.length >= 8 && data.subarray(0, 8).equals(PNG_SIGNATURE)) {
    dims = parsePng(data);
  } else if (data.length >= 2 && data[0] === 0xff && data[1] === 0xd8) {
    dims = parseJpeg(data);
  } else {
    throw new ImageError("unsupported_image", "not a PNG or JPEG header");
  }
  checkDims(dims);
  return dims;
}

function checkDims(dims: ImageDims): void {
  if (dims.width <= 0 || dims.height <= 0) {
    throw new ImageError("image_bad_dimensions", "non-positive image dimension");
  }
  if (dims.width > MAX_IMAGE_WIDTH || dims.height > MAX_IMAGE_HEIGHT) {
    throw new ImageError(
      "image_too_large",
      `${dims.width}x${dims.height} exceeds ${MAX_IMAGE_WIDTH}x${MAX_IMAGE_HEIGHT}`,
    );
  }
  if (dims.width * dims.height > MAX_TOTAL_PIXELS) {
    throw new ImageError(
      "image_too_large",
      `${dims.width * dims.height} pixels exceeds total budget ${MAX_TOTAL_PIXELS}`,
    );
  }
}

function parsePng(data: Buffer): ImageDims {
  if (data.length < 24 || data.subarray(12, 16).toString("latin1") !== "IHDR") {
    throw new ImageError("image_malformed", "PNG missing IHDR chunk");
  }
  return { width: data.readUInt32BE(16), height: data.readUInt32BE(20), kind: "png" };
}

function parseJpeg(data: Buffer): ImageDims {
  let i = 2;
  const n = data.length;
  while (i + 1 < n) {
    if (data[i] !== 0xff) {
      i++;
      continue;
    }
    const marker = data[i + 1] as number;
    i += 2;
    if (marker === 0xd8 || marker === 0xd9 || (marker >= 0xd0 && marker <= 0xd7)) {
      continue;
    }
    if (i + 2 > n) break;
    const segLen = data.readUInt16BE(i);
    if (segLen < 2) throw new ImageError("image_malformed", "JPEG segment length underflow");
    if (JPEG_SOF.has(marker)) {
      if (i + 7 > n) throw new ImageError("image_malformed", "JPEG SOF truncated");
      return { width: data.readUInt16BE(i + 5), height: data.readUInt16BE(i + 3), kind: "jpeg" };
    }
    i += segLen;
  }
  throw new ImageError("image_malformed", "no JPEG SOF marker found");
}

File: server/src/hephaestus/agent_bridge/limits.py
"""Bridge limits: the single source of truth loaded from ``schemas/bridge_limits.json``.

Every architecture §5 numeric limit lives in that JSON file; this module loads it
verbatim and exposes typed accessors plus the bounded validators the bridge runs
*before* trusting a payload:

* :func:`validate_json_structure` — depth / members / array / string budgets applied
  to an already-parsed value (defence against pathological but well-formed JSON).
* :func:`enforce_max_utf8_bytes` — exact UTF-8 sizing with unpaired-surrogate
  rejection (``invalid_unicode_scalar``), used for ``x-hephaestus-maxUtf8Bytes``.
* :func:`parse_image_header` — a bounded PNG/JPEG header parser that recovers the
  declared dimensions *before* any full decode, rejecting decompression bombs.

No limit literal is duplicated here — read them off :data:`LIMITS`.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Final, cast

from hephaestus.core.limits import limits_path as core_limits_path

__all__ = [
    "LIMITS",
    "TURN_SECONDS",
    "ImageDims",
    "ImageError",
    "LimitError",
    "enforce_binary_budget",
    "enforce_max_utf8_bytes",
    "limits_path",
    "parse_image_header",
    "validate_json_structure",
]


def limits_path() -> Path:
    """Absolute path to the loaded limits file (for provenance/logging).

    Delegates to :func:`hephaestus.core.limits.limits_path`. This module used to
    carry its own copy of the walk-up search, which meant two resolvers for one
    file — and both were wrong in an installed wheel, where the walk climbs out
    of ``site-packages`` and finds nothing. ``hephaestus-server`` already depends
    on ``hephaestus-core``, so the core resolver (which knows about the packaged
    copy the wheel ships) is the one that should answer for both.
    """
    return core_limits_path()


def _load() -> dict[str, Any]:
    with limits_path().open("r", encoding="utf-8") as fh:
        data: dict[str, Any] = json.load(fh)
    return data


LIMITS: Final[dict[str, Any]] = _load()

# Convenience typed views over the loaded document (no literal duplication).
_WIRE: Final[dict[str, Any]] = LIMITS["wire"]
_JSON: Final[dict[str, Any]] = LIMITS["json"]
_IMAGE: Final[dict[str, Any]] = LIMITS["image"]
_BINARY: Final[dict[str, Any]] = LIMITS["binary"]
_HTTP: Final[dict[str, Any]] = LIMITS["http"]

FRAME_VERSION: Final[int] = int(_WIRE["frame_version"])
MAX_FRAME_BYTES: Final[int] = int(_WIRE["max_frame_bytes"])

#: The ceiling on ONE HTTP request body, enforced by ``hephaestus.http.app``
#: from ``Content-Length`` and again while the body streams (``INTERFACE.md``
#: §2.4, audit-2026-09-04 J-http-limits-1). Exported beside the frame cap and
#: deliberately far below it: the frame cap bounds what the *bridge transport*
#: can carry between two trusted processes, while this bounds what an untrusted
#: client can make the serving process allocate. The largest legitimate body on
#: the HTTP surface is a part script; a prompt is capped an order of magnitude
#: smaller again (:data:`PROMPT_MAX_UTF8_BYTES`), so a megabyte leaves every
#: real body room and still refuses the 64 MiB one that used to be buffered
#: whole before anything looked at it.
MAX_REQUEST_BYTES: Final[int] = int(_HTTP["max_request_bytes"])
MAX_JSON_DEPTH: Final[int] = int(_JSON["max_depth"])
MAX_JSON_MEMBERS: Final[int] = int(_JSON["max_members"])
MAX_JSON_ARRAY_ITEMS: Final[int] = int(_JSON["max_array_items"])
MAX_STRING_BYTES: Final[int] = int(_JSON["max_string_bytes"])
MAX_BINARY_BYTES: Final[int] = int(_BINARY["max_binary_bytes"])

#: One model TURN's budget, in seconds — the bound on a ``session.prompt`` call.
#:
#: A turn is not a tool. ``timeouts.tool_seconds`` bounds ONE tool call, and a
#: turn runs a model round trip plus however many tools the model asks for, one
#: of which may be a build the sidecar itself allows ``cad_build_seconds``
#: (300 s). Defaulting the turn call to the tool bound therefore placed a
#: 120 s ceiling around a thing that legitimately takes longer, and the
#: watchdog's remedy for an overdue call is to kill the whole sidecar — every
#: session in the process — over latency that was never a fault. Set to the
#: delegated child's own default (``delegation.deadline_default_seconds``)
#: because a delegated child turn IS a turn: the two must not disagree about
#: how long one is allowed to take.
TURN_SECONDS: Final[float] = float(LIMITS["timeouts"]["turn_seconds"])
MAX_IMAGE_BYTES: Final[int] = int(_IMAGE["max_image_bytes"])
MAX_IMAGE_WIDTH: Final[int] = int(_IMAGE["max_width"])
MAX_IMAGE_HEIGHT: Final[int] = int(_IMAGE["max_height"])
MAX_TOTAL_PIXELS: Final[int] = int(_IMAGE["max_total_pixels"])
MAX_IMAGES_PER_RESULT: Final[int] = int(_IMAGE["max_images_per_result"])
MAX_PENDING_RPC: Final[int] = int(LIMITS["rpc"]["max_pending"])
PROMPT_MAX_UTF8_BYTES: Final[int] = int(LIMITS["prompt"]["max_utf8_bytes"])


class LimitError(ValueError):
    """A payload exceeded a structural or size limit. ``code`` is stable."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


class ImageError(LimitError):
    """An image header was malformed or its dimensions exceeded the budget."""


@dataclass(frozen=True)
class ImageDims:
    """Dimensions recovered from an image header before any full decode."""

    width: int
    height: int
    kind: str  # "png" | "jpeg"

    @property
    def pixels(self) -> int:
        return self.width * self.height


def validate_json_structure(value: object) -> None:
    """Reject well-formed-but-pathological JSON against the §5 structural caps.

    Applies, over the whole parsed value: nesting ``depth <= MAX_JSON_DEPTH``,
    per-object ``members <= MAX_JSON_MEMBERS``, per-array ``items <=
    MAX_JSON_ARRAY_ITEMS``, and per-string ``utf-8 bytes <= MAX_STRING_BYTES``.
    Raises :class:`LimitError` on the first violation.
    """

    # Iterative walk to avoid Python recursion limits on adversarial nesting.
    stack: list[tuple[object, int]] = [(value, 1)]
    while stack:
        node, depth = stack.pop()
        if depth > MAX_JSON_DEPTH:
            raise LimitError("json_too_deep", f"nesting exceeds depth {MAX_JSON_DEPTH}")
        if isinstance(node, dict):
            node_dict = cast("dict[str, object]", node)
            if len(node_dict) > MAX_JSON_MEMBERS:
                raise LimitError(
                    "json_too_many_members",
                    f"object has {len(node_dict)} members (max {MAX_JSON_MEMBERS})",
                )
            for key, child in node_dict.items():
                _check_string(key)
                stack.append((child, depth + 1))
        elif isinstance(node, list):
            node_list = cast("list[object]", node)
            if len(node_list) > MAX_JSON_ARRAY_ITEMS:
                raise LimitError(
                    "json_array_too_long",
                    f"array has {len(node_list)} items (max {MAX_JSON_ARRAY_ITEMS})",
                )
            for child in node_list:
                stack.append((child, depth + 1))
        elif isinstance(node, str):
            _check_string(node)


def _check_string(s: str) -> None:
    # A surrogate-bearing string cannot be encoded as UTF-8; reject it as a
    # structural violation rather than letting encode() raise opaquely later.
    n = _utf8_len_strict(s)
    if n > MAX_STRING_BYTES:
        raise LimitError(
            "json_string_too_large",
            f"string is {n} bytes (max {MAX_STRING_BYTES})",
        )


def _utf8_len_strict(s: str) -> int:
    """UTF-8 byte length, rejecting any unpaired UTF-16 surrogate scalar.

    JSON permits ``\\uD800`` escapes that decode to lone surrogates in Python;
    the bridge treats those as ``invalid_unicode_scalar`` before sizing or
    hashing, matching the TypeScript/MCP rule (no replacement-character coercion).
    """
    for ch in s:
        o = ord(ch)
        if 0xD800 <= o <= 0xDFFF:
            raise LimitError(
                "invalid_unicode_scalar",
                f"unpaired UTF-16 surrogate U+{o:04X} is not a valid scalar",
            )
    return len(s.encode("utf-8"))


def enforce_binary_budget(total_bytes: int, *, field: str = "result") -> None:
    """Enforce the AGGREGATE binary budget of one tool result (J-http-limits-9).

    ``binary.max_binary_bytes`` was exported on both sides of the bridge and
    validated by nothing — a grep over the whole tree returned this export, the
    TypeScript export, and the test that pinned it as dead. It has exactly one
    defensible subject here: the only binary payloads that cross this bridge are
    the base64 image blocks of a tool result, and ``max_image_bytes`` (8 MiB)
    times ``max_images_per_result`` (4) is precisely this cap, so the intended
    meaning is the per-result *aggregate*. Nothing summed them, so a result with
    four maximal images passed four per-image checks and no aggregate one.

    The mirror is ``agent/src/limits.ts``'s ``enforceBinaryBudget``, called from
    the sidecar's image loop; this side is enforced where a result's image
    blocks are assembled, because a limit enforced on one side of this bridge is
    not enforced.
    """
    if total_bytes > MAX_BINARY_BYTES:
        raise LimitError(
            "binary_too_large",
            f"{field} carries {total_bytes} binary bytes (max {MAX_BINARY_BYTES})",
        )


def enforce_max_utf8_bytes(value: str, max_bytes: int, *, field: str = "value") -> int:
    """Enforce ``x-hephaestus-maxUtf8Bytes``; return the exact UTF-8 byte length.

    Rejects unpaired surrogates first (``invalid_unicode_scalar``), then sizes
    the exact UTF-8 encoding. Never truncates. Raises :class:`LimitError` with
    code ``prompt_too_large`` when ``max_bytes`` is exceeded.
    """
    n = _utf8_len_strict(value)
    if n > max_bytes:
        raise LimitError(
            "prompt_too_large",
            f"{field} is {n} UTF-8 bytes (max {max_bytes})",
        )
    return n


def parse_image_header(data: bytes) -> ImageDims:
    """Recover image dimensions from a PNG or JPEG header without full decode.

    Enforces the per-image byte budget, then reads only the header fields needed
    for width/height and checks them against the dimension and total-pixel
    budgets. Raises :class:`ImageError` on malformed data or a bomb; the full
    decoder still enforces its own allocation cap downstream.
    """
    if len(data) > MAX_IMAGE_BYTES:
        raise ImageError(
            "image_too_large",
            f"image is {len(data)} bytes (max {MAX_IMAGE_BYTES})",
        )
    if data[:8] == b"\x89PNG\r\n\x1a\n":
        dims = _parse_png(data)
    elif data[:2] == b"\xff\xd8":
        dims = _parse_jpeg(data)
    else:
        raise ImageError("unsupported_image", "not a PNG or JPEG header")
    _check_dims(dims)
    return dims


def _check_dims(dims: ImageDims) -> None:
    if dims.width <= 0 or dims.height <= 0:
        raise ImageError("image_bad_dimensions", "non-positive image dimension")
    if dims.width > MAX_IMAGE_WIDTH or dims.height > MAX_IMAGE_HEIGHT:
        raise ImageError(
            "image_too_large",
            f"{dims.width}x{dims.height} exceeds {MAX_IMAGE_WIDTH}x{MAX_IMAGE_HEIGHT}",
        )
    if dims.pixels > MAX_TOTAL_PIXELS:
        raise ImageError(
            "image_too_large",
            f"{dims.pixels} pixels exceeds total budget {MAX_TOTAL_PIXELS}",
        )


def _parse_png(data: bytes) -> ImageDims:
    # Signature (8) + IHDR length (4) + "IHDR" (4) + width (4 BE) + height (4 BE).
    if len(data) < 24 or data[12:16] != b"IHDR":
        raise ImageError("image_malformed", "PNG missing IHDR chunk")
    width = int.from_bytes(data[16:20], "big")
    height = int.from_bytes(data[20:24], "big")
    return ImageDims(width=width, height=height, kind="png")


# JPEG start-of-frame markers that carry dimensions (exclude DHT/JPG/DAC/RST/…).
_JPEG_SOF = frozenset(
    {0xC0, 0xC1, 0xC2, 0xC3, 0xC5, 0xC6, 0xC7, 0xC9, 0xCA, 0xCB, 0xCD, 0xCE, 0xCF}
)


def _parse_jpeg(data: bytes) -> ImageDims:
    # Walk marker segments from just after SOI (FF D8) until an SOF marker.
    i = 2
    n = len(data)
    while i + 1 < n:
        if data[i] != 0xFF:
            # Skip fill bytes / stray data until the next marker prefix.
            i += 1
            continue
        marker = data[i + 1]
        i += 2
        if marker in (0xD8, 0xD9) or 0xD0 <= marker <= 0xD7:
            # SOI/EOI/RSTn carry no length payload.
            continue
        if i + 2 > n:
            break
        seg_len = int.from_bytes(data[i : i + 2], "big")
        if seg_len < 2:
            raise ImageError("image_malformed", "JPEG segment length underflow")
        if marker in _JPEG_SOF:
            # length(2) + precision(1) + height(2) + width(2)
            if i + 7 > n:
                raise ImageError("image_malformed", "JPEG SOF truncated")
            height = int.from_bytes(data[i + 3 : i + 5], "big")
            width = int.from_bytes(data[i + 5 : i + 7], "big")
            return ImageDims(width=width, height=height, kind="jpeg")
        i += seg_len
    raise ImageError("image_malformed", "no JPEG SOF marker found")
